File: jscomp/test/test_bug.js
'use strict';

var Bytes = require("../../lib/js/bytes.js");
var Test_char = require("./test_char.js");
var Caml_bytes = require("../../lib/js/caml_bytes.js");

function escaped(s) {
  var n = 0;
  for(var i = 0 ,i_finish = s.length; i < i_finish; ++i){
    var c = s[i];
    var tmp;
    var exit = 0;
    if (c >= 14) {
      if (c !== 34 && c !== 92) {
        exit = 1;
      } else {
        tmp = 2;
      }
    } else if (c >= 11) {
      if (c >= 13) {
        tmp = 2;
      } else {
        exit = 1;
      }
    } else if (c >= 8) {
      tmp = 2;
    } else {
      exit = 1;
    }
    if (exit === 1) {
      tmp = Test_char.caml_is_printable(c) ? 1 : 4;
    }
    n = n + tmp | 0;
  }
  if (n === s.length) {
    return Bytes.copy(s);
  }
  var s$p = Caml_bytes.caml_create_bytes(n);
  n = 0;
  for(var i$1 = 0 ,i_finish$1 = s.length; i$1 < i_finish$1; ++i$1){
    var c$1 = s[i$1];
    var exit$1 = 0;
    if (c$1 > 92 || c$1 < 34) {
      if (c$1 >= 14) {
        exit$1 = 1;
      } else {
        switch (c$1) {
          case 8 :
              s$p[n] = /* '\\' */92;
              n = n + 1 | 0;
              s$p[n] = /* 'b' */98;
              break;
          case 9 :
              s$p[n] = /* '\\' */92;
              n = n + 1 | 0;
              s$p[n] = /* 't' */116;
              break;
          case 10 :
              s$p[n] = /* '\\' */92;
              n = n + 1 | 0;
              s$p[n] = /* 'n' */110;
              break;
          case 0 :
          case 1 :
          case 2 :
          case 3 :
          case 4 :
          case 5 :
          case 6 :
          case 7 :
          case 11 :
          case 12 :
              exit$1 = 1;
              break;
          case 13 :
              s$p[n] = /* '\\' */92;
              n = n + 1 | 0;
              s$p[n] = /* 'r' */114;
              break;
          
        }
      }
    } else if (c$1 > 91 || c$1 < 35) {
      s$p[n] = /* '\\' */92;
      n = n + 1 | 0;
      s$p[n] = c$1;
    } else {
      exit$1 = 1;
    }
    if (exit$1 === 1) {
      if (Test_char.caml_is_printable(c$1)) {
        s$p[n] = c$1;
      } else {
        s$p[n] = /* '\\' */92;
        n = n + 1 | 0;
        s$p[n] = 48 + (c$1 / 100 | 0) | 0;
        n = n + 1 | 0;
        s$p[n] = 48 + (c$1 / 10 | 0) % 10 | 0;
        n = n + 1 | 0;
        s$p[n] = 48 + c$1 % 10 | 0;
      }
    }
    n = n + 1 | 0;
  }
  return s$p;
}

exports.escaped = escaped;
/* No side effect */

File: jscomp/test/test_bug.ml
external string_length : string -> int = "%string_length"
external unsafe_blit_string : string -> int -> bytes -> int -> int -> unit
                     = "caml_blit_string" "noalloc"

external char_code: char -> int = "%identity"
external char_chr: int -> char = "%identity"
open Bytes
let escaped s =
  let n = ref 0 in
  for i = 0 to length s - 1 do
    n := !n +
      (match unsafe_get s i with
       | '"' | '\\' | '\n' | '\t' | '\r' | '\b' -> 2
       | c -> if Test_char.caml_is_printable c then 1 else 4)
  done;
  if !n = length s then copy s else begin
    let s' = create !n in
    n := 0;
    for i = 0 to length s - 1 do
      begin match unsafe_get s i with
      | ('"' | '\\') as c ->
          unsafe_set s' !n '\\'; incr n; unsafe_set s' !n c
      | '\n' ->
          unsafe_set s' !n '\\'; incr n; unsafe_set s' !n 'n'
      | '\t' ->
          unsafe_set s' !n '\\'; incr n; unsafe_set s' !n 't'
      | '\r' ->
          unsafe_set s' !n '\\'; incr n; unsafe_set s' !n 'r'
      | '\b' ->
          unsafe_set s' !n '\\'; incr n; unsafe_set s' !n 'b'
      | c ->
          if Test_char.caml_is_printable c then
            unsafe_set s' !n c
          else begin
            let a = char_code c in
            unsafe_set s' !n '\\';
            incr n;
            unsafe_set s' !n (char_chr (48 + a / 100));
            incr n;
            unsafe_set s' !n (char_chr (48 + (a / 10) mod 10));
            incr n;
            unsafe_set s' !n (char_chr (48 + a mod 10))
          end
      end;
      incr n
    done;
    s'
  end
